src/lib/profiles/data-management/Current/LoggingManagement.cpp

```diff
@@ -2146,6 +2246,161 @@ void CircularEventReader::Init(CircularEventBuffer * inBuf)
+    SuccessOrExit(err);
+
+#if WEAVE_CONFIG_EVENT_LOGGING_UTC_TIMESTAMPS
+    err = writer.Put(ContextTag(kTag_PersistEvent_FirstEventUTCTimestamp), mFirstEventUTCTimestamp);
+    SuccessOrExit(err);
+
+    err = writer.Put(ContextTag(kTag_PersistEvent_LastEventUTCTimestamp), mLastEventUTCTimestamp);
+    SuccessOrExit(err);
+
+    err = writer.PutBoolean(ContextTag(kTag_PersistEvent_UTCInitialized), mUTCInitialized);
+    SuccessOrExit(err);
+#endif
+
+    err = writer.EndContainer(container);
+    SuccessOrExit(err);
+
+exit:
+    if (err != WEAVE_NO_ERROR)
+    {
+        WeaveLogError(EventLogging, "Serialize event error: %d", err);
+    }
+    return err;
+}
+
+WEAVE_ERROR CircularEventBuffer::LoadEvents(TLVReader & reader)
+{
+    WEAVE_ERROR err = WEAVE_NO_ERROR;
+
+    TLVType container;
+
+    uint32_t counterValue;
+    uint32_t startingCounterValue;
+    uint8_t importance;
+
+    err = reader.Next(kTLVType_Structure, AnonymousTag);
+    SuccessOrExit(err);
+    err = reader.EnterContainer(container);
+    SuccessOrExit(err);
+
+    err = reader.Next(kTLVType_UnsignedInteger, ContextTag(kTag_PersistEvent_ImportanceLevel));
+    SuccessOrExit(err);
+    err = reader.Get(importance);
+    SuccessOrExit(err);
+    mImportance = (ImportanceType)importance;
+
+    err = reader.Next(kTLVType_ByteString, ContextTag(kTag_PersistEvent_EventData));
+    SuccessOrExit(err);
+    VerifyOrExit(reader.GetLength() <= mBuffer.GetQueueSize(), err = WEAVE_ERROR_BUFFER_TOO_SMALL);
+    mBuffer.SetQueueLength(reader.GetLength());
+    mBuffer.SetQueueHead(mBuffer.GetQueue());
+    err = reader.GetBytes(mBuffer.GetQueue(), mBuffer.DataLength());
+    SuccessOrExit(err);
+
+    err = reader.Next(kTLVType_UnsignedInteger, ContextTag(kTag_PersistEvent_FirstEventId));
+    SuccessOrExit(err);
+    err = reader.Get(mFirstEventID);
+    SuccessOrExit(err);
+
+    err = reader.Next(kTLVType_UnsignedInteger, ContextTag(kTag_PersistEvent_LastEventId));
+    SuccessOrExit(err);
+    err = reader.Get(mLastEventID);
+    SuccessOrExit(err);
+
+    err = reader.Next(kTLVType_UnsignedInteger, ContextTag(kTag_PersistEvent_FirstEventTimestamp));
+    SuccessOrExit(err);
+    err = reader.Get(mFirstEventTimestamp);
+    SuccessOrExit(err);
+
+    err = reader.Next(kTLVType_UnsignedInteger, ContextTag(kTag_PersistEvent_LastEventTimestamp));
+    SuccessOrExit(err);
+    err = reader.Get(mLastEventTimestamp);
+    SuccessOrExit(err);
+
+    err = reader.Next(kTLVType_UnsignedInteger, ContextTag(kTag_PersistEvent_EventIdCounter));
+    SuccessOrExit(err);
+    err = reader.Get(counterValue);
+    SuccessOrExit(err);
+
+    err = reader.Next(kTLVType_UnsignedInteger, ContextTag(kTag_PersistEvent_StartingCounterValue));
+    SuccessOrExit(err);
+    err = reader.Get(startingCounterValue);
+    SuccessOrExit(err);
+
+    static_cast<PersistedCounter *>(mEventIdCounter)->ResumeEpoch(startingCounterValue, counterValue);
```

Comment:
Why is this necessary?   The event id counter is already persistent.   Overwriting the counter value when LoadEvents() is called could possibly make the counter value roll backwards, which seems incorrect.

Reply:
This is to fix the mistake I made earlier. When the device wakes up, event counter starts from a new epoch, and event buffer would have events counters from two epochs, which might spread a wide range. The problem is - When going through events in the event buffer for uploading, it always starts from the lowest counter in the buffer and +1 in each iteration... Then the weave thread would be block...

Reply:
@jaylogue , if we are persisting events and loading was successful, then does the persistedEventIdCounter need to be used? It should help only in the case of a crash or when the persisting/loading failed, isn't it? Otherwise, it could, potentially, start after the lastEventID?